### PR TITLE
Fix ABA problem of lock-free read by re-calculate checksum

### PR DESCRIPTION
**Pull request by JiayuZzz**
Re-calculate read data checksum, and compare with PMem checksum, if fail then redo read

### CODE REVIEW

**Review by KVKIT — APPROVED** [review]

File: engine/kv_engine.cpp [hunk 1]
```diff
@@ -623,56 +623,50 @@ Status KVEngine::Recovery() {
 
 Status KVEngine::HashGetImpl(const pmem::obj::string_view &key,
                              std::string *value, uint16_t type_mask) {
-  std::unique_ptr<DataEntry> data_entry(
+  std::unique_ptr<DataEntry> data_entry_meta(
```
Comment (ZiyanShi) [inline engine/kv_engine.cpp:626]: No need to use polymorphism if we just use Header and Meta when Hash Get.

File: tests/tests.cpp [hunk 2]
```diff
@@ -38,7 +38,7 @@ class EngineBasicTest : public testing::Test {
     random_str(&str_pool[0], str_pool_length);
     configs.pmem_file_size = (16ULL << 30);
     configs.populate_pmem_space = false;
-    configs.hash_bucket_num = (1 << 5);
+    configs.hash_bucket_num = (1 << 10);
     configs.hash_bucket_size = 64;
```
Comment (ZiyanShi) [inline tests/tests.cpp:42]: What's this two configs?
We should let user decide hash entries per bucket and hash buckets per slot, but not the hash_bucket_size

**Review by ZiyanShi — APPROVED** [review]

File: tests/tests.cpp (hunk 2 above)
Comment (JiayuZzz) [reply tests/tests.cpp:42]: agree

File: engine/kv_engine.cpp (hunk 1 above)
Comment (JiayuZzz) [reply engine/kv_engine.cpp:626]: hash table search will copy DLDataentry if it's a sorted record

File: tests/tests.cpp (hunk 2 above)
Comment (JiayuZzz) [reply tests/tests.cpp:42]: but use size, it's clear to align bucket size to cache line

**Pull request merged**